gfx.h

```diff
@@ -191,15 +191,23 @@ GfxRaytracingPrimitive const *gfxAccelerationStructureGetRaytracingPrimitives(Gf
 
 class GfxRaytracingPrimitive { GFX_INTERNAL_NAMED_HANDLE(GfxRaytracingPrimitive); public: };
 
+enum GfxRaytracingPrimitiveFlags
+{
+    GfxRaytracingPrimitiveFlags_None = 0,
+    GfxRaytracingPrimitiveFlags_NonOpaque = 1,
+
+    GfxRaytracingPrimitiveFlags_Count
+};
+
 GfxRaytracingPrimitive gfxCreateRaytracingPrimitive(GfxContext context, GfxAccelerationStructure acceleration_structure);
 GfxResult gfxDestroyRaytracingPrimitive(GfxContext context, GfxRaytracingPrimitive raytracing_primitive);
 
-GfxResult gfxRaytracingPrimitiveBuild(GfxContext context, GfxRaytracingPrimitive raytracing_primitive, GfxBuffer vertex_buffer, uint32_t vertex_stride = 0);
-GfxResult gfxRaytracingPrimitiveBuild(GfxContext context, GfxRaytracingPrimitive raytracing_primitive, GfxBuffer index_buffer, GfxBuffer vertex_buffer, uint32_t vertex_stride = 0);
+GfxResult gfxRaytracingPrimitiveBuild(GfxContext context, GfxRaytracingPrimitive raytracing_primitive, GfxBuffer vertex_buffer, uint32_t vertex_stride = 0, GfxRaytracingPrimitiveFlags flags = GfxRaytracingPrimitiveFlags_NonOpaque);
+GfxResult gfxRaytracingPrimitiveBuild(GfxContext context, GfxRaytracingPrimitive raytracing_primitive, GfxBuffer index_buffer, GfxBuffer vertex_buffer, uint32_t vertex_stride = 0, GfxRaytracingPrimitiveFlags flags = GfxRaytracingPrimitiveFlags_NonOpaque);
 GfxResult gfxRaytracingPrimitiveSetTransform(GfxContext context, GfxRaytracingPrimitive raytracing_primitive, float const *row_major_4x4_transform);
 GfxResult gfxRaytracingPrimitiveSetInstanceID(GfxContext context, GfxRaytracingPrimitive raytracing_primitive, uint32_t instance_id);   // retrieved through `ray_query.CommittedInstanceID()`
 GfxResult gfxRaytracingPrimitiveSetInstanceMask(GfxContext context, GfxRaytracingPrimitive raytracing_primitive, uint8_t instance_mask);
-GfxResult gfxRaytracingPrimitiveUpdate(GfxContext context, GfxRaytracingPrimitive raytracing_primitive);
+GfxResult gfxRaytracingPrimitiveUpdate(GfxContext context, GfxRaytracingPrimitive raytracing_primitive, GfxRaytracingPrimitiveFlags flags = GfxRaytracingPrimitiveFlags_NonOpaque);
```

Comment:
Do we really need to pass the flags again on update?
I would have thought having them on build should suffice; do you think it's beneficial to be able to change those on update?

Reply:
The only reason thats there is because `updateRaytracingPrimitive` calls `buildRaytracingPrimitive` which requires the flag parameter be passed. Should the update fail then it creates a new one which requires that flag. So I wasnt sure what the likelihood of the update failing (if ever?) and building a new one. If it never fails then the flag can be ignored.

Reply:
Yeah, BVH updates in DXR are achieved through calling the build function indeed, but passing in the `D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAG_PERFORM_UPDATE` flag.

So you'll most likely need to persist the flags passed by the user on build into the `RaytracingPrimitive` structure: https://github.com/gboisse/gfx/blob/master/gfx.h#L776-L788

I'm unsure what you mean with the update failing, but yes safe to assume it doesn't here I think 🙂
D3D12 being explicit, the only thing that can fail on an update would be outside the `BuildRaytracingAccelerationStructure()` and along the lines of failing to allocate enough memory for the BVH, the scratch, etc.
However, sufficient memory should already have been allocated on the initial build.

Reply:
I did see the `RaytracingPrimitive` struct and was considering that but I was waiting for feedback. Ive added the flags to that in the most recent commit.